feat: add minter role to TestERC20

just-mitch opened the pull request:
Allows multiple accounts to be minters of the staking and fee assets.

Creates a FeeAssetHandler in the periphery to allow mints of a fixed size.

Also allows rollup owner to update the mana target (and thus, the mana limit which is twice the target).

See [design](https://github.com/AztecProtocol/engineering-designs/blob/42455c99b867cde4d67700bc97ac12309c2332ea/docs/faucets/dd.md#testerc20)

Fix #12887
Fix #12882

LHerskind commented on l1-contracts/src/core/RollupCore.sol:
```diff
@@ -126,6 +126,10 @@ contract RollupCore is
   }
 
+  function updateManaTarget(uint256 _manaTarget) external override(ITestRollup) onlyOwner {
+    FeeLib.updateManaTarget(_manaTarget);
```
We should have an event here. We have been pretty bad generally at adding events when state change, but it is usually good practice since it makes analysis and such much simpler.

LHerskind reviewed (commented):
Mainly the ownerships that need to be addressed such that we don't just give everyone with any rights admin rights.

just-mitch commented on l1-contracts/src/mock/TestERC20.sol:
```diff
@@ -7,27 +7,41 @@ import {ERC20} from "@oz/token/ERC20/ERC20.sol";
-  // solhint-disable-next-line comprehensive-interface
-  function setFreeForAll(bool _freeForAll) external onlyOwner {
-    freeForAll = _freeForAll;
+  function removeMinter(address _minter) external override(IMintableERC20) onlyMinter {
```
since this is `onlyOwner` now, it is safe to allow owner to relinquish their minting. I didn't want someone else to remove the owner as a minter.

just-mitch committed:
Allows multiple accounts to be minters of the staking and fee assets.

Creates a FeeAssetHandler in the periphery to allow mints of a fixed size.

Also allows rollup owner to update the mana target (and thus, the mana limit which is twice the target).

LHerskind commented on l1-contracts/src/mock/TestERC20.sol:
```diff
@@ -7,27 +7,46 @@ import {ERC20} from "@oz/token/ERC20/ERC20.sol";
-    if (msg.sender != owner() && !freeForAll) {
-      revert("Not owner or free for all");
-    }
+  event MinterAdded(address indexed minter);
```
I think the linter would be yelling at you here, but maybe not 🤔

just-mitch replied:
no it isn't. would you expect state variable below events and errors?

LHerskind replied:
Oh, no actually https://docs.soliditylang.org/en/latest/style-guide.html#order-of-layout you are right. Think just not used to seeing them as they are usually in the interface. Would probably move them in there actually, such that just with the `IMintableERC20` you have what you need to understand the errors.

LHerskind commented on l1-contracts/src/mock/TestERC20.sol:
```diff
@@ -7,27 +7,46 @@ import {ERC20} from "@oz/token/ERC20/ERC20.sol";
+    }
+    minters[owner()] = false;
+    _transferOwnership(newOwner);
+    minters[newOwner] = true;
```
Slightly annoying here, but this is not emitting the events that you would probably expect, simplest might be to make the add and remove public and call those directly.

LHerskind commented on l1-contracts/src/mock/TestERC20.sol:
```diff
@@ -7,27 +7,46 @@ import {ERC20} from "@oz/token/ERC20/ERC20.sol";
-  function setFreeForAll(bool _freeForAll) external onlyOwner {
-    freeForAll = _freeForAll;
+  {
+    minters[_owner] = true;
```
Could also use the `addMinter` here to get the event 🤷

just-mitch replied:
But then the caller is the contract, and not _owner?

LHerskind commented on l1-contracts/test/faucet/FeeAssetHandler.t.sol:
```diff
@@ -0,0 +1,45 @@
+    assertEq(testERC20.balanceOf(_recipient), feeAssetHandler.mintAmount());
+  }
+
+  function setUp() public {
```
Usually you would write the setup at the top so it is easier to read through.

LHerskind commented on l1-contracts/test/faucet/FeeAssetHandler.t.sol:
```diff
@@ -0,0 +1,45 @@
+    vm.assume(_recipient != address(this));
+    vm.assume(_recipient != address(0));
+    feeAssetHandler.mint(_recipient);
+    assertEq(testERC20.balanceOf(_recipient), feeAssetHandler.mintAmount());
```
The caller is not changed here, so it don't make sense to be `WhenAnyoneCallsMint` it is only the recipient that is changed here. Should change the caller with a prank as well.

just-mitch replied:
my mistake! Thank you

LHerskind commented on l1-contracts/test/TestERC20.t.sol:
```diff
@@ -1,24 +1,90 @@
+    vm.prank(address(_minter));
+    testERC20.mint(address(this), 100);
+
+    vm.expectEmit();
```
You need to update the `expectEmit` to ensure that it is actually checking that the values in the event makes sense and that it is from the correct address.

As I recall, the first 3 bools are for the indexed values, then for the data and then the address that should emit.

```solidity
vm.expectEmit(true, true, true, true, address(testERC20));
```

LHerskind commented on l1-contracts/test/TestERC20.t.sol:
```diff
@@ -1,24 +1,90 @@
+    vm.assume(_minter != address(0));
+
+    vm.prank(address(_minter));
+    vm.expectRevert();
```
Should check the revert message as well. Also, would prefer that this was not just done in here, would make more sense as a separate test, similar to the `test_mint`.

LHerskind commented on l1-contracts/test/TestERC20.t.sol:
```diff
@@ -1,24 +1,90 @@
+
+  function test_onlyOwnerCanAddRemoveMinter(address _minter, address _caller) public {
+    vm.assume(_minter != address(this));
+    vm.assume(_minter != address(0));
```
Not sure about why the address 0 case?

LHerskind commented on l1-contracts/test/TestERC20.t.sol:
```diff
@@ -1,24 +1,90 @@
+    vm.assume(_minter != address(this));
+    vm.assume(_minter != address(0));
+
+    vm.assume(_caller != address(_minter));
```
Not sure about these as well, only the owner should be able to add, so we don't need to constrain on these as well.

LHerskind commented on l1-contracts/test/TestERC20.t.sol:
```diff
@@ -1,24 +1,90 @@
+
+    address oldOwner = testERC20.owner();
+
+    vm.prank(_newOwner);
```
Generally for the expectRevert we should be expecting an explicit error.

LHerskind commented on l1-contracts/test/faucet/FeeAssetHandler.tree:
```diff
@@ -0,0 +1,7 @@
+FeeAssetHandlerTest
```
Would say that the tree is slightly misused here, or maybe just used differently that what I would. 
I would probably do it for just one function instead, and then see the branches more clearly defined. Take the branches that revert first so we can have all of those dealt with and use the modifiers to read what goes on deeper down.

```
# when caller is not owner
## it reverts
# when caller is owner
## it updates the mint amount
## it emit {MintAmountUpdated} event
```

LHerskind commented on spartan/aztec-network/files/config/config-full-node-env.sh:
```diff
@@ -7,6 +7,7 @@ if [ -n "$REGISTRY_CONTRACT_ADDRESS" ] && [ -n "$BOOTSTRAP_NODES" ]; then
 export BOOTSTRAP_NODES=$BOOTSTRAP_NODES
 export REGISTRY_CONTRACT_ADDRESS=$REGISTRY_CONTRACT_ADDRESS
 export SLASH_FACTORY_CONTRACT_ADDRESS=$SLASH_FACTORY_CONTRACT_ADDRESS
+export FEE_ASSET_HANDLER_CONTRACT_ADDRESS=$FEE_ASSET_HANDLER_CONTRACT_ADDRESS
```
Dear lord almighty, save me from the env vars

![Bruce Almighty I Got The Power GIF by Jim Carrey](https://media0.giphy.com/media/fwta5c1e0O2hymy9p3/giphy.gif)

LHerskind commented on yarn-project/bot/src/factory.ts:
```diff
@@ -101,7 +101,7 @@ export class BotFactory {
       this.log.info(`Deploying account at ${address}`);
 
-      const claim = await this.bridgeL1FeeJuice(address, 10n ** 22n);
+      const claim = await this.bridgeL1FeeJuice(address, MINT_AMOUNT);
```
Would we need it to mint itself a couple of times, or do we get sufficient to cover our fees for some time? Just thinking that it would be annoying if we run into the issue where at block 4000 it runs out of juice.

just-mitch commented on yarn-project/ethereum/src/contracts/fee_asset_handler.test.ts:
```diff
@@ -0,0 +1,83 @@
+
+const originalVersionSalt = 42;
+
+describe('Registry', () => {
```
copy pasta

just-mitch commented on yarn-project/ethereum/src/contracts/fee_asset_handler.ts:
```diff
@@ -0,0 +1,39 @@
+// See https://github.com/AztecProtocol/engineering-designs/blob/42455c99b867cde4d67700bc97ac12309c2332ea/docs/faucets/dd.md#choosing-the-mint-amount
+export const MINT_AMOUNT = 1_000_000_000_000_000n;
+
+export class FeeAssetHandlerContract {
```
Yeah agreed.

just-mitch committed:
begin addressing feedback

just-mitch committed:
rewriting tests

just-mitch committed:
add tests

just-mitch committed:
delete old test

just-mitch committed:
less assumptions

just-mitch committed:
update tests
increase base mint

just-mitch committed:
clean up tests

just-mitch committed:
fix build and gas report

just-mitch committed:
fix tests

just-mitch committed:
fix more tests
